Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/auto/AutoSubmission.kt

```diff
@@ -0,0 +1,144 @@
+import androidx.work.ExistingPeriodicWorkPolicy
+import androidx.work.NetworkType
+import androidx.work.PeriodicWorkRequestBuilder
+import androidx.work.WorkManager
+import de.rki.coronawarnapp.submission.SubmissionSettings
+import de.rki.coronawarnapp.submission.task.SubmissionTask
+import de.rki.coronawarnapp.task.TaskController
+import de.rki.coronawarnapp.task.TaskInfo
+import de.rki.coronawarnapp.task.common.DefaultTaskRequest
+import de.rki.coronawarnapp.task.submitBlocking
+import de.rki.coronawarnapp.util.TimeStamper
+import de.rki.coronawarnapp.worker.BackgroundConstants
+import kotlinx.coroutines.flow.map
+import org.joda.time.Instant
+import timber.log.Timber
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class AutoSubmission @Inject constructor(
+    private val timeStamper: TimeStamper,
+    private val submissionSettings: SubmissionSettings,
+    private val workManager: WorkManager,
+    private val taskController: TaskController
+) {
+
+    private var currentMode: Mode = Mode.DISABLED
+
+    private fun List<TaskInfo>.isSubmissionTaskRunning() = any {
+        it.taskState.isActive && it.taskState.request.type == SubmissionTask::class
+    }
+
+    val isSubmissionRunning = taskController.tasks.map { it.isSubmissionTaskRunning() }
+
+    fun setup() {
+        Timber.tag(TAG).v("setup()")
+
+        if (submissionSettings.autoSubmissionEnabled.value) {
+            Timber.tag(TAG).i("Fresh app start and auto submission is enabled, updating mode.")
+
+            updateMode(Mode.SUBMIT_ASAP)
+        } else {
+            Timber.tag(TAG).d("AutoSubmission is disabled.")
+        }
+    }
+
+    fun updateMode(newMode: Mode) {
+        Timber.tag(TAG).i("updateMode(mode=$newMode)")
+
+        when (newMode) {
+            Mode.DISABLED -> disableAutoSubmission()
+            Mode.MONITOR -> enableAutoSubmission(lastActivity = timeStamper.nowUTC)
+            Mode.SUBMIT_ASAP -> enableAutoSubmission(lastActivity = Instant.EPOCH)
+        }
+        currentMode = newMode
+    }
+
+    suspend fun runSubmissionNow() {
+        Timber.tag(TAG).i("runSubmissionNow()")
+
+        val result = taskController.submitBlocking(
+            DefaultTaskRequest(
+                type = SubmissionTask::class,
+                arguments = SubmissionTask.Arguments(checkUserActivity = false),
+                originTag = TAG
+            )
+        )
+        if (result.isSuccessful) {
+            Timber.tag(TAG).i("Blocking submission was successful.")
+        } else {
+            Timber.tag(TAG).w(result.error, "Blocking submission was not successful, enabling auto submission.")
+            updateMode(Mode.SUBMIT_ASAP)
+        }
+    }
+
+    private fun scheduleWorker() {
+        Timber.tag(TAG).v("scheduleWorker(): Creating periodic worker request for submission.")
+
+        val request = PeriodicWorkRequestBuilder<SubmissionWorker>(15, TimeUnit.MINUTES).apply {
```

Comment:
15 Minutes...
Do you think it could affect battery consumption?

Reply:
Probably, but it's a short task. And there is really no alternative. I'd actually like to have it quicker, but 15 is the minimum. If we increase this, we raise the chance that the user does something to prevent us from submitting the keys ever.

e.g. get a call during submission, app in background, app is killed, user thinks "positive test :frowning:", forgets about the submission, maybe even uninstalls the app...

Reply:
The worker is only scheduled though once there is positive test result, so until that, there should be no impact.